### DISCUSSION TITLE
Each scenario on a new thread.

### DISCUSSION THREAD

**Discussion by zie1ony** [cucumber-rs/cucumber#286] [Q&A] [answered]
I have a code that uses `thread_local!`. It requires each scenario to run on a new thread, exactly like a standard behaviour of a standard `#[test]` tagged tests. Is it possible to force cucumber to do that?

**Comment by ilslv** [comment]
@zie1ony unfortunately we don't support this feature and have no plans to do it, as we targeting runtime-agnostic async test, which doesn't play nicely with spawning threads on each test. But in the book we briefly describe a way to [implement a custom `Runner`](https://cucumber-rs.github.io/cucumber/main/architecture/runner.html). If you'll decide to go with this approach I will happily help.

**Reply by tyranron** [reply]
@zie1ony also, nothing prevents you from decorating your tests which do use that `thread_local!` to run in a separate thread via `thread::spawn` or similar.

**Reply by zie1ony** [reply]
@ilslv I'll give it a try.
@tyranron how would that look like?

**Reply by tyranron** [reply]
@zie1ony the idea is to have something like the following:
```rust
// Helper function for decoration.
async fn in_separate_thread<F: FnOnce() + Send>(f: F) {
    // We use async channel here, to not block the current thread on
    // await the spawned thread to finish.
    let (tx, rx) = futures::channel::oneshot::channel();
    std::thread::spawn(move || {
        // Additionally, we can start here single-threaded runtime, if we need to
        // use async inside `f`.

        f();

        // Additionally, we shoud `.catch_unwind` any panics produces by `f`
        // return them via oneshot channel.

        tx.send(()).unwrap();
    })
    rx.await;
}

#[then("whatever")]
async fn my_step(w: &mut World) {
      in_separate_thread(|| {
         // step logic here
      }).await
}
```

However, this would run each step in its own thread, not the whole scenario altogether.  Which may be not what you want.

Still, this idea may be extended to cover all the scenario: we spawn a separate thread on `World` initialization (or lazily, on demand, but store thread handle in `World`), and instead of spawning each time a new thread, our decorator will send a function to be executed onto that already spawned thread in each step. This way, the same `thread_local!` could be preserved between steps of scenario.

**Reply by zie1ony** [reply]
@tyranron that's a bit complicated. This probably will force my code to change a lot.
@ilslv I worked a bit with custom runner. I managed to send a single scenario to new thread but the problem is getting results back. 

I think that the issue is the `Info` from `events.rs` that cannot be sharde between threads. I'm getting this error:
```
`(dyn std::any::Any + std::marker::Send + 'static)` cannot be shared between threads safely
```
Would you consider making `Info` something more thread friendly?

**Reply by zie1ony** [reply]
I have followed this thread: https://users.rust-lang.org/t/return-value-from-catch-unwind-is-a-useless-any/89134/5.
They say to downcast errors to `String`, so I have __blindly__ changed `Info` to `String` in `events.rs` and then treated all errors with
```rust
fn downcast_to_string(v: Box<dyn Any + Send>) -> String {
    match v.downcast_ref::<&'static str>() {
        Some(s) => String::from(*s),
        None => match v.downcast_ref::<String>() {
            Some(s) => s.clone(),
            None => String::from("Box<dyn Any>"),
        },
    }
}
```
That actually worked and now `execute_scenario` from `CustomRunner` is:
```rust
 async fn execute_scenario(
      scenario: gherkin::Scenario,
  ) -> impl Stream<Item = event::Feature<MyWorld>> {
      let s = scenario.clone();
      let (tx, rx) = futures::channel::oneshot::channel();
      thread::spawn(move || {
          let hook = panic::take_hook();
          panic::set_hook(Box::new(|_| {}));
          
          let steps = block_on(async {
              let mut steps = Vec::with_capacity(s.steps.len());
              let mut world = MyWorld::new().await.unwrap();
              for step in s.steps.clone() {
                  let (w, ev) = Self::execute_step(world, step.clone()).await;
                  world = w;
                  let should_stop = matches!(ev, event::Step::Failed(..));
                  steps.push((step, ev));
                  if should_stop {
                      break;
                  }
              }
              steps
          }); 
          
          panic::set_hook(hook);
          tx.send(steps).unwrap();
      });

      let steps = rx.await.unwrap();

      let scenario = Arc::new(scenario);
      stream::once(future::ready(event::Scenario::Started))
          .chain(stream::iter(steps.into_iter().flat_map(|(step, ev)| {
              let step = Arc::new(step);
              [
                  event::Scenario::Step(step.clone(), event::Step::Started),
                  event::Scenario::Step(step, ev),
              ]
          })))
          .chain(stream::once(future::ready(event::Scenario::Finished)))
          .map(move |event| event::Feature::Scenario(
              scenario.clone(), 
              event::RetryableScenario { event, retries: None },
          ))
  }
 ```

This doesn't work well with panics (yet). It's just proof of concept. Would you help me going this direction?

**Reply by ilslv** [reply]
@zie1ony 

> Would you consider making Info something more thread friendly?

The thing is, rust allows you to panic not only with `String`s,  but with [arbitrary types too](https://doc.rust-lang.org/stable/std/panic/fn.panic_any.html) which can be `!Sync` for example. I would prefer to give users more options rather than immediately downcasting `dyn Any + Send` to `String`s. 

> Would you help me going this direction?

Proof of concept looks fine. Do you have concrete questions about pushing it forward?

**Reply by zie1ony** [reply]
The question is still the same: How to send `cucumber::event::Step` to another thread?

**Reply by ilslv** [reply, marked as answer]
@zie1ony as I said earlier, to give users maximum flexibility, I would prefer to keep `event::Step: !Sync`, which makes it impossible to pass value to another thread. I would suggest implementing your own stricter version of `event::Step`, which will look something like:

```rust
#[derive(Debug)]
pub enum SyncStep<World> {
    Started,
    Skipped,
    Passed(regex::CaptureLocations, Option<step::Location>),
    Failed(
        Option<regex::CaptureLocations>,
        Option<step::Location>,
        Option<Arc<World>>,
        StepError,
    ),
}

#[derive(Clone, Debug, Display, Error, From)]
pub enum SyncStepError {
    #[display(fmt = "Step doesn't match any function")]
    NotFound,
    #[display(fmt = "Step match is ambiguous: {}", _0)]
    AmbiguousMatch(step::AmbiguousMatchError),
    #[display(fmt = "Step panicked. Captured output: {}", "coerce_error(_0)")]
    Panic(#[error(not(source))] String),
    //        Instead of `Info` ^^^^^^
}
```

and implement conversions from/to `event::Step`. So you will convert `event::Step` into a `event::SyncStep`, pass it to another thread and then convert back to `event::Step`

**Reply by zie1ony** [reply]
Thank you @ilslv
